nifi-nar-bundles/nifi-apicurio-bundle/nifi-apicurio-schema-registry-service/src/main/java/org/apache/nifi/apicurio/schemaregistry/client/CachingSchemaRegistryClient.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.apicurio.schemaregistry.client;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.nifi.serialization.record.RecordSchema;
+
+import java.time.Duration;
+
+public class CachingSchemaRegistryClient implements SchemaRegistryClient {
+    private final SchemaRegistryClient client;
+    private final LoadingCache<String, RecordSchema> nameCache;
+    private final LoadingCache<Pair<String, Integer>, RecordSchema> nameVersionCache;
```

Comment:
Seems a bit confusing to have a single `Cache Size` property but then have two caches - would it be better to combine these together, passing the `OptionalInt` from the `ApicuriSchemaRegistry` then using a default value (e.g. `-1` that wouldn't be valid as input to the service normally) to keep everything in a single cache, e.g. `nameVersionCache` - would that make more sense, or do we have precendent of multiple caches elsewhere already?

Reply:
Thanks for the suggestion. We have precedents in RestSchemaRegistry and GlueSchemaRegistry but I think this makes sense and viable.